File: platform/script-debugger/debugger-ui/src/com/jetbrains/javascript/debugger/NameMapper.java
/*
 * Copyright 2000-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.javascript.debugger;

import com.google.common.base.CharMatcher;
import com.intellij.openapi.editor.Document;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import gnu.trove.THashMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.debugger.sourcemap.MappingEntry;
import org.jetbrains.debugger.sourcemap.MappingList;
import org.jetbrains.debugger.sourcemap.SourceMap;

import java.util.Map;

import static org.jetbrains.rpc.CommandProcessor.LOG;

public class NameMapper {
  public static final String S1 = ",()[]{}=";
  private static final CharMatcher NAME_TRIMMER = CharMatcher.INVISIBLE.or(CharMatcher.anyOf(S1 + ".&:"));
  // don't trim trailing .&: - could be part of expression
  private static final CharMatcher OPERATOR_TRIMMER = CharMatcher.INVISIBLE.or(CharMatcher.anyOf(S1));

  private final Document document;
  private final Document generatedDocument;
  private final MappingList sourceMappings;
  private final SourceMap sourceMap;

  private Map<String, String> rawNameToSource;

  public NameMapper(@NotNull Document document, @NotNull Document generatedDocument, @NotNull MappingList sourceMappings, @NotNull SourceMap sourceMap) {
    this.document = document;
    this.generatedDocument = generatedDocument;
    this.sourceMappings = sourceMappings;
    this.sourceMap = sourceMap;
  }

  @Nullable
  public Map<String, String> getRawNameToSource() {
    return rawNameToSource;
  }

  // PsiNamedElement, JSVariable for example
  // returns generated name
  @Nullable
  public String map(@NotNull PsiElement identifierOrNamedElement) {
    int offset = identifierOrNamedElement.getTextOffset();
    int line = document.getLineNumber(offset);

    int sourceEntryIndex = sourceMappings.indexOf(line, offset - document.getLineStartOffset(line));
    if (sourceEntryIndex == -1) {
      return null;
    }

    MappingEntry sourceEntry = sourceMappings.get(sourceEntryIndex);
    MappingEntry next = sourceMappings.getNextOnTheSameLine(sourceEntryIndex, false);
    if (next != null && sourceMappings.getColumn(next) == sourceMappings.getColumn(sourceEntry)) {
      warnSeveralMapping(identifierOrNamedElement);
      return null;
    }

    String sourceEntryName = sourceEntry.getName();
    String generatedName = extractName(getGeneratedName(generatedDocument, sourceMap, sourceEntry), true);
    if (!generatedName.isEmpty()) {
      String sourceName = sourceEntryName;
      if (sourceName == null) {
        sourceName = identifierOrNamedElement instanceof PsiNamedElement ? ((PsiNamedElement)identifierOrNamedElement).getName() : identifierOrNamedElement.getText();
        if (sourceName == null) {
          return null;
        }
      }

      if (rawNameToSource == null) {
        rawNameToSource = new THashMap<String, String>();
      }
      rawNameToSource.put(generatedName, sourceName);
      return generatedName;
    }
    return null;
  }

  public static void warnSeveralMapping(@NotNull PsiElement element) {
    // see https://dl.dropboxusercontent.com/u/43511007/s/Screen%20Shot%202015-01-21%20at%2020.33.44.png
    // var1 mapped to the whole "var c, notes, templates, ..." expression text + unrelated text "   ;"
    LOG.warn("incorrect sourcemap, several mappings for named element " + element.getText());
  }

  @NotNull
  public static String trimName(@NotNull CharSequence rawGeneratedName, boolean isLastToken) {
    return (isLastToken ? NAME_TRIMMER : OPERATOR_TRIMMER).trimFrom(rawGeneratedName);
  }

  @NotNull
  protected String extractName(@NotNull CharSequence rawGeneratedName, boolean isLastToken) {
    return trimName(rawGeneratedName, isLastToken);
  }

  @NotNull
  private static CharSequence getGeneratedName(@NotNull Document document, @NotNull SourceMap sourceMap, @NotNull MappingEntry sourceEntry) {
    int lineStartOffset = document.getLineStartOffset(sourceEntry.getGeneratedLine());
    MappingEntry nextGeneratedMapping = sourceMap.getMappings().getNextOnTheSameLine(sourceEntry);
    int endOffset;
    if (nextGeneratedMapping == null) {
      endOffset = document.getLineEndOffset(sourceEntry.getGeneratedLine());
    }
    else {
      endOffset = lineStartOffset + nextGeneratedMapping.getGeneratedColumn();
    }
    return document.getImmutableCharSequence().subSequence(lineStartOffset + sourceEntry.getGeneratedColumn(), endOffset);
  }
}

File: platform/script-debugger/debugger-ui/src/com/intellij/javascript/debugger/NameMapper.kt
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.javascript.debugger

import com.google.common.base.CharMatcher
import com.intellij.openapi.editor.Document
import com.intellij.openapi.util.TextRange
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.SyntaxTraverser
import org.jetbrains.debugger.sourcemap.MappingEntry
import org.jetbrains.debugger.sourcemap.Mappings
import org.jetbrains.debugger.sourcemap.MappingsProcessorInLine
import org.jetbrains.debugger.sourcemap.SourceMap
import org.jetbrains.rpc.LOG
import java.util.concurrent.atomic.AtomicInteger

private const val S1 = ",()[]{}="
// don't trim trailing .&: - could be part of expression
private val OPERATOR_TRIMMER = CharMatcher.invisible().or(CharMatcher.anyOf(S1))

val NAME_TRIMMER: CharMatcher = CharMatcher.invisible().or(CharMatcher.anyOf("$S1.&:"))

// generateVirtualFile only for debug purposes
open class NameMapper(private val document: Document, private val transpiledDocument: Document, private val sourceMappings: Mappings, protected val sourceMap: SourceMap, private val transpiledFile: VirtualFile? = null) {
  var rawNameToSource: MutableMap<String, String>? = null
    private set

  // PsiNamedElement, JSVariable for example
  // returns generated name
  @JvmOverloads
  open fun map(identifierOrNamedElement: PsiElement, saveMapping: Boolean = true): String? {
    return doMap(identifierOrNamedElement, false, saveMapping)
  }

  protected fun doMap(identifierOrNamedElement: PsiElement, mapBySourceCode: Boolean, saveMapping: Boolean = true): String? {
    val mappings = getMappingsForElement(identifierOrNamedElement)
    if (mappings == null || mappings.isEmpty()) return null
    val sourceEntry = mappings[0]
    val generatedName: String?
    try {
      generatedName = extractName(getGeneratedName(identifierOrNamedElement, transpiledDocument, sourceMap, mappings), identifierOrNamedElement)
    }
    catch (e: IndexOutOfBoundsException) {
      LOG.warn("Cannot get generated name: source entry (${sourceEntry.generatedLine},  ${sourceEntry.generatedColumn}). Transpiled File: " + transpiledFile?.path)
      return null
    }
    if (generatedName == null || generatedName.isEmpty()) {
      return null
    }

    var sourceName = sourceEntry.name
    if (sourceName == null || mapBySourceCode) {
      sourceName = (identifierOrNamedElement as? PsiNamedElement)?.name ?: identifierOrNamedElement.text ?: sourceName ?: return null
    }

    if (saveMapping) {
      addMapping(generatedName, sourceName)
    }
    return generatedName
  }

  protected open fun getMappingsForElement(element: PsiElement): List<MappingEntry>? {
    val mappings = mutableListOf<MappingEntry>()

    val offset = element.textOffset
    val line = document.getLineNumber(offset)
    val elementColumn = offset - document.getLineStartOffset(line)
    val elementEndColumn = elementColumn + element.textLength - 1

    val sourceEntryIndex = sourceMappings.indexOf(line, elementColumn)
    if (sourceEntryIndex == -1) {
      return null
    }

    val sourceEntry = sourceMappings.getByIndex(sourceEntryIndex)
    val next = sourceMappings.getNextOnTheSameLine(sourceEntryIndex, false)
    if (next != null && sourceMappings.getColumn(next) == sourceMappings.getColumn(sourceEntry)) {
      warnSeveralMapping(element)
      return null
    }

    val file = element.containingFile
    val namedElementsCounter = AtomicInteger(0)
    val processor = object : MappingsProcessorInLine {
      override fun process(entry: MappingEntry, nextEntry: MappingEntry?): Boolean {
        val entryColumn = entry.sourceColumn
        // next entry column could be equal to prev, see https://code.google.com/p/google-web-toolkit/issues/detail?id=9103
        val isSuitable = if (nextEntry == null || (entryColumn == 0 && nextEntry.sourceColumn == 0)) {
          entryColumn <= elementColumn
        }
        else {
          entryColumn in elementColumn..elementEndColumn
        }
        if (isSuitable) {
          val startOffset = document.getLineStartOffset(line) + entryColumn
          val endOffset =
            if (nextEntry != null) document.getLineStartOffset(line) + nextEntry.sourceColumn
            else document.getLineEndOffset(line)
          if (collectNamedElementsInRange(TextRange(startOffset, endOffset))) {
            return false
          }

          mappings.add(entry)
        }
        return true
      }

      private fun collectNamedElementsInRange(range: TextRange): Boolean {
        return SyntaxTraverser.psiTraverser(file)
          .onRange(range)
          .filter { it is PsiNamedElement && range.contains(it.textRange) && namedElementsCounter.incrementAndGet() > 1 }
          .traverse()
          .isNotEmpty
      }
    }
    if (!sourceMap.processSourceMappingsInLine(sourceEntry.source, sourceEntry.sourceLine, processor)) {
      return null
    }
    return mappings
  }

  fun addMapping(generatedName: String, sourceName: String) {
    if (rawNameToSource == null) {
      rawNameToSource = HashMap<String, String>()
    }
    rawNameToSource!!.put(generatedName, sourceName)
  }

  protected open fun extractName(rawGeneratedName: CharSequence?, context: PsiElement? = null):String? = rawGeneratedName?.let {
    NAME_TRIMMER.trimFrom(it)
  }

  companion object {
    fun trimName(rawGeneratedName: CharSequence, isLastToken: Boolean): String? = (if (isLastToken) NAME_TRIMMER else OPERATOR_TRIMMER).trimFrom(rawGeneratedName)
  }

  protected open fun getGeneratedName(element: PsiElement, document: Document, sourceMap: SourceMap, mappings: List<MappingEntry>): CharSequence? {
    val sourceEntry = mappings[0]
    val lineStartOffset = document.getLineStartOffset(sourceEntry.generatedLine)
    val nextGeneratedMapping = sourceMap.generatedMappings.getNextOnTheSameLine(sourceEntry)
    val endOffset: Int
    if (nextGeneratedMapping == null) {
      endOffset = document.getLineEndOffset(sourceEntry.generatedLine)
    }
    else {
      endOffset = lineStartOffset + nextGeneratedMapping.generatedColumn
    }
    return document.immutableCharSequence.subSequence(lineStartOffset + sourceEntry.generatedColumn, endOffset)
  }
}

fun warnSeveralMapping(element: PsiElement) {
  // see https://dl.dropboxusercontent.com/u/43511007/s/Screen%20Shot%202015-01-21%20at%2020.33.44.png
  // var1 mapped to the whole "var c, notes, templates, ..." expression text + unrelated text "   ;"
  LOG.warn("incorrect sourcemap, several mappings for named element ${element.text}")
}

